Doc/extending/newtypes_tutorial.rst

```diff
@@ -0,0 +1,865 @@
+   Traceback (most recent call last):
+     File "<stdin>", line 1, in <module>
+   TypeError: cannot add type "noddy.Noddy" to string
+
+Note that the name is a dotted name that includes both the module name and the
+name of the type within the module. The module in this case is :mod:`noddy` and
+the type is :class:`Noddy`, so we set the type name to :class:`noddy.Noddy`.
+Using the real dotted import path is important to make your type compatible
+with the :mod:`pydoc` and :mod:`pickle` modules. ::
+
+   .tp_basicsize = sizeof(NoddyObject),
+   .tp_itemsize = 0,
+
+This is so that Python knows how much memory to allocate when creating
+new :class:`Noddy` instances.  :c:member:`~PyTypeObject.tp_itemsize` is
+only used for variable-sized objects and should otherwise be zero.
+
+.. note::
+
+   If you want your type to be subclassable from Python, and your type has the same
+   :c:member:`~PyTypeObject.tp_basicsize` as its base type, you may have problems with multiple
+   inheritance.  A Python subclass of your type will have to list your type first
+   in its :attr:`~class.__bases__`, or else it will not be able to call your type's
+   :meth:`__new__` method without getting an error.  You can avoid this problem by
+   ensuring that your type has a larger value for :c:member:`~PyTypeObject.tp_basicsize` than its
+   base type does.  Most of the time, this will be true anyway, because either your
+   base type will be :class:`object`, or else you will be adding data members to
+   your base type, and therefore increasing its size.
+
+We set the class flags to :const:`Py_TPFLAGS_DEFAULT`. ::
+
+   .tp_flags = Py_TPFLAGS_DEFAULT,
+
+All types should include this constant in their flags.  It enables all of the
+members defined until at least Python 3.3.  If you need further members,
+you will need to OR the corresponding flags.
+
+We provide a doc string for the type in :c:member:`~PyTypeObject.tp_doc`. ::
+
+   .tp_doc = "Noddy objects",
+
+To enable object creation, we have to provide a :c:member:`~PyTypeObject.tp_new`
+implementation. In this case, we can just use the default implementation
+provided by the API function :c:func:`PyType_GenericNew`. ::
+
+   .tp_new = PyType_GenericNew,
+
+Everything else in the file should be familiar, except for some code in
+:c:func:`PyInit_noddy`::
+
+   if (PyType_Ready(&NoddyType) < 0)
+       return;
+
+This initializes the :class:`Noddy` type, filling in a number of members
+to the appropriate default values, including :attr:`ob_type` that we initially
+set to *NULL*. ::
+
+   PyModule_AddObject(m, "Noddy", (PyObject *) &NoddyType);
+
+This adds the type to the module dictionary.  This allows us to create
+:class:`Noddy` instances by calling the :class:`Noddy` class::
+
+   >>> import noddy
+   >>> mynoddy = noddy.Noddy()
+
+That's it!  All that remains is to build it; put the above code in a file called
+:file:`noddy.c` and ::
+
+   from distutils.core import setup, Extension
```

Comment:
Using `setuptools` is likely to be more robust here, since it tends to be better at finding compilers, emits up to date package metadata, and can be used to build wheel archives in combination with `bdist_wheel`.

https://packaging.python.org/tutorials/installing-packages/#ensure-pip-setuptools-and-wheel-are-up-to-date is the appropriate reference if folks just want to build locally, while they'll want https://packaging.python.org/tutorials/distributing-packages/ if they plan to actually create an sdist and/or wheel file for their extension module.

Reply:
Right, I should add a link to that external resource.